### ISSUE TITLE
Use a real automation/config management tool

### ISSUE THREAD

**Issue by ageis** [nextcloud/vm#951]
Hey, this is just a suggestion, so feel free to close it after reading.

I appreciate your work a lot, but I think it's a little bit silly that you're automating everything with Bash scripts. I do note that you're incorporating Vagrant and Terraform...

I took a look over the hundreds of issues in this repository, and I have the feeling a lot of the bugs, and many of your constraints regarding OS distros and VMs, and the myriad edge cases that arise, might be avoidable. 

In my humble opinion, in the future the Bash scripts here should be converted to [Ansible](https://github.com/ansible/ansible) roles and playbooks. If it's not in your wheelhouse yet then I recommend [learning it if you can](https://docs.ansible.com/ansible/latest/index.html). It'll take care of idempotency for you and make things that you do in shell scripts a lot easier by abstracting a lot of stuff away... Especially if Nextcloud is supposed to be state of the art. Right now it is fairly difficult and confusing to deploy, and some of that's because of the scripts.

Don't get me wrong, bash scripts for automation are a tried and true staple which will never be irrelevant. But IMO it's a dated and inferior pattern compared to things like Ansible (and Salt, Chef, Puppet).

I'd be happy to do the development and conversion myself if it were a paid job... Anyway, best of luck.

**Comment by enoch85** [comment]
Thanks for your suggestions. 

Thank you for your suggestions.

Ansible is already old. :) There are new languages all the time, and converting would essentially mean a rewrite of everything.

> if it were a paid job

Yeah, that's [the problem](https://onezero.medium.com/the-internet-relies-on-people-working-for-free-a79104a68bcc?source=friends_link&sk=c9416c806a8ad2b8fca89074f94454ac). 

I don't get paid to do this either. My time is limited (father + full-time job + company) and I sometimes wish the days had 48 hours.

I'll leave this open in case anyone wants to rewrite it.

**Comment by enoch85** [comment]
Thank you for your suggestions.

Ansible is already old. :) There are new languages all the time, and converting would essentially mean a rewrite of everything.

> if it were a paid job

Yeah, that's [the problem](https://onezero.medium.com/the-internet-relies-on-people-working-for-free-a79104a68bcc?source=friends_link&sk=c9416c806a8ad2b8fca89074f94454ac). 

I don't get paid to do this either. My time is limited (father + full-time job + company) and I sometimes wish the days had 48 hours.

I'll leave this open in case anyone wants to rewrite it.

**Comment by enoch85** [comment]
@ageis 

Btw:

> Right now it is fairly difficult and confusing to deploy, and some of that's because of the scripts.

I don't know what you think is difficult with launching one single script and just follow the instructions? Please explain.

**Comment by enoch85** [comment]
@ageis Here's a more detailed version of my last post.

> I appreciate your work a lot, but I think it's a little bit silly...

Thanks? :confused: 

I have more than 5 years invested in this project, doing everything in the open-source fashion (free) for people around the world to enjoy. You come here, ranting about how much doing it in bash repo sucks basically, saying that you want to get paid to help out - saying that I shouldn't get you wrong. Well... 

Instead of ranting, send a PR! Enjoy the feeling knowing how many people that are using your stuff. Thousands! It's a great feeling making the world a better place (making it easier to use Nextcloud).

To get things going it might be a good idea to migrate one of the scripts to Ansible. I just had a quick look (10 minutes) and I don't see though how you can automate stuff with `whiptails`, `if commands` and such, in the same way that the bash scripts are built now. It feels like Ansible would be good to do the basic stuff, but the if logic would be handled by bash, correct? I mean what about the [if](https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L524-L544)? Would a mix of bash and Ansible be the correct way forward, because Ansible can't do certain stuff that bash can?

> Vagrant and Terraform...

Yeah, as a way of building the VM for the maintainers (me), not the end user. It would make it easier to do releases. 

> bash scripts for automation are a tried and true staple which will never be irrelevant.

You got that one right! :smile: :1st_place_medal: 

These scripts are used by thousands of people if not more, and have to take into account a lot of stuff. Like, if this VM was installed 4 years ago, certain dependencies were used, and things were handled differently. The update script must be backwards compatible. Would Ansible be able to handle that?

> and I have the feeling a lot of the bugs, and many of your constraints regarding **OS distros and VMs**

I actually don't know what you mean by that. It's run on Ubuntu, and Ubuntu only, that's always been the case. So please explain what you mean by "bugs with "OS distros", because there's only one distro it can be run on. Regarding VMs, all supported Hypervisors are listed in the [docs](https://docs.hanssonit.se). Among them are VMware, VirtualBox, Hyper-V, and yeah, essentially every Hypervisor where you can install Ubuntu, plenty. :)

> fairly difficult and confusing to deploy

Agreed, it was some years ago, but not anymore. :) Coming from you with your (at least alleged Github Description) experience, that statement makes me surprised... Based on [previous experience](https://www.techandme.se/feedback/) (discontinued in 2018), that someone that never used Linux before (or knows what a VM is) can do it with a successful result, then in *my humble** option, I don't agree with you on that.

As you don't seem to be updated on the subject, here are some examples of how you can deploy Nextcloud:
- With Docker
- With Ansible (yes it exits)
- As a pre-configrued Virtual Machine
- As a snap
- Through good old fashion bash scripts
- With a PHP installer
- ++ More ways I can't think of now. 

> some of that's because of the scripts. 

Hmm... So you think it's hard to [run a script that installs Nextcloud and just follow the given instructions](https://www.youtube.com/watch?v=ceOld6P16wI)? But yeah, I don't know Ansible, maybe it can do magic?

**TLDR:**
I agree that Bash are well proven. I agree that Ansible (or an even better tool) might be a good way forward. "Don't get me wrong..." I have nothing against improving. Ansible is a good tool, sure. These scripts works perfectly fine though, so my motivation of changing isn't on the highest level so to speak.

It just feels like you actually aren't sticking to facts regarding the current state of this repo (or Nextcloud for that matter) which makes your statements totally wrong. Try the scripts, actually read the issues in this repo, look around for ways of installing Nextcloud, and then tell me again what you think. 

You are free to open a PR to get things going, I'd be happy if you would. I don't mind improving stuff and make it easier to maintain.
